ERCS/eip-send-tx-conditional.md

```diff
@@ -0,0 +1,154 @@
+---
+eip: send-tx-cond
+title: Conditional send transaction RPC
+description: JSON-RPC API for block builders allowing users to express preconditions for transaction inclusion
+author: Dror Tirosh (@drortirosh), Yoav Weiss (@yoavw), Alex Forshtat (@forshtat), Shahaf Nacson (@shahafn)
+discussions-to:
+status: Draft
+type: Standards Track
+category: Interface
+created: 2024-04-16
+---
+
+## Abstract
+
+This EIP proposes a new JSON-RPC API method `eth_sendRawTransactionConditional` for block builders and sequencers,
+enhancing transaction integration by allowing users to express preconditions for transaction inclusion.
+
+This method aims to improve efficiency by reducing the need for transaction simulation,
+thereby improving the computational efficiency of transaction ordering.
+
+## Motivation
+
+Current private block builder APIs, such as the Flashbots API,
+require block builders to simulate transactions to determine eligibility for inclusion,
+a process that is CPU-intensive and inefficient.
+
+The proposed RPC method addresses this by enabling transactions to specify preconditions,
+thus reducing computational overhead and potentially lowering transaction costs.
+
+Moreover, the flashbots API does not provide the block builder with a mechanism to determine the
+cross-dependencies of different transactions.
+
+The only way to guarantee that another transaction does not interfere with a given one is by placing
+it as the first transaction in the block.
+This makes this placement very lucrative, and disproportionately expensive.
+
+In addition, since there is no way to give any guarantee on other slots, their pricing has to be low accordingly.
+
+Since there is no easy way to detect cross-dependencies of different transactions,
+it is CPU-intensive to find an optimal ordering of transactions.
+
+## Specification
+
+* Method: `eth_sendRawTransactionConditional`
+
+* Parameters:
+
+1. `transaction`: The raw, signed transaction data. Similar to `eth_sendRawTransaction`.
+2. `options`: An object containing conditions under which the transaction must be included.
+* The `options` parameter may include any of the following optional members:
+    * **knownAccounts**: a mapping of accounts with their expected storage slots' values.
+        * The key of the mapping is account address.
+        * A special key `balance` defines the expected balance of the account.
```

Comment:
```suggestion
        * A special key `balance` defines the expected balance of the account.
        * A special key `code` to expect specific code: `""` to mark "expected not to have code", and `"0xef01…"` for specific eip-7702 delegation. (Other values are irrelevant, since code other code values can't be modified)
```

Reply:
also need to add `nonce`. it is critical if the transaction contains 7702 tuples, and we need to know if they will be applied or not (and with that, we cover all contract state info: code, balance, nonce, so there are not more added fields)

Reply:
RE: `Other values are irrelevant, since code other code values can't be modified`
Removing this comment for two reasons:
1. It may be used to express a statement "only send a transaction after this contract has been deployed"
2. There are tricks with CREATE2 and SELFDESTRUCT that enable modifiable contract code. This is not widely used but it is technically possible (until SELFDESTRUCT is completely removed from all chains)

Reply:
Added nonce.